File: hw2/sender.py
import socket
import struct
import sys
import select
import hashlib

agentip = '127.0.0.1'
agentport = 8888
ip = '127.0.0.1'
port = 8887


class Header:
	def __init__(self, length, seqNumber, ackNumber, fin, syn, ack):
		self.length = length
		self.seqNumber = seqNumber
		self.ackNumber = ackNumber
		self.fin = fin
		self.syn = syn
		self.ack = ack

class segment:
	def __init__(self, header, data):
		self.header = header
		self.data = data


emptydata = ''
for i in range(1000):
	emptydata += '\0'
sendersocket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sendersocket.bind((ip,port))
filename = sys.argv[1]
f = open(filename,'rb')
seqnum = 1
winseqnum = seqnum
maxseqnum = seqnum
endseqnum = 0
winsend = 0
timeout = 0
getacknum = 1
threshold = 16
winsize = 1
drop = 0
reads = [sendersocket]
while True:
	winseqnum = seqnum
	for i in range(winsize):
		change = 0
		f.seek(1000*(seqnum-1), 0)
		data = f.read(1000)
		header = Header(1000,seqnum,0,0,0,0)
		if 0 <= len(data) < 1000:
			if len(data) == 0:
				endseqnum = seqnum
				header.length = 0
				break
			for i in range(len(data),1000):
				header.length = len(data)
				data = data
		Message = segment(header, data)
		Message = struct.pack("6i1000s",Message.header.length, Message.header.seqNumber, Message.header.ackNumber, Message.header.fin, Message.header.syn, Message.header.ack, Message.data)
		sendersocket.sendto(Message,(agentip, agentport))
		seqnum += 1
		if seqnum > maxseqnum:
			maxseqnum = seqnum
			change = 1
		if change == 1:
			print('send    data    #',seqnum - 1,',    winsize =', winsize)
		else:
			print('resnd   data    #',seqnum - 1,',    winsize =',winsize)

	while winseqnum != seqnum:
		timeout = 0
		readable, writeable, exceptional = select.select(reads,[],[],1)
		if readable != []:
			for read in readable:
				ackdata, addr = sendersocket.recvfrom(10240)
				if addr[0] == agentip and addr[1] == agentport:
					length, seqNumber, ackNumber, fin, syn, ack, data = struct.unpack("6i1000s",ackdata)
					print('recv    ack     #',ackNumber)
					if ackNumber == winseqnum:
						winseqnum += 1
		else:
			threshold = max(int(winsize/2),1)
			winsize = 1
			timeout = 1
			seqnum = winseqnum
			print('time    out,             threshold = ',threshold)
			break
	if timeout == 0:
		if winsize < threshold:
			winsize = winsize*2
		else:
			winsize += 1
	if winseqnum == endseqnum:
		header = Header(0,0,0,1,0,0)
		Message = struct.pack("6i1000s",0,0,0,1,0,0,emptydata.encode('utf-8'))
		sendersocket.sendto(Message,(agentip, agentport))
		print('send    fin')
		ackdata, addr = sendersocket.recvfrom(10240)
		if addr[0] == agentip and addr[1] == agentport:
			length, seqNumber, ackNumber, fin, syn, ack, data = struct.unpack("6i1000s",ackdata)
			if fin == 1 and ack == 1:
				print('recv    finack')
				break 

File: hw2/sender.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <string.h>

typedef struct {
	int length;
	int seqNumber;
	int ackNumber;
	int fin;
	int syn;
	int ack;
} header;

typedef struct{
	header head;
	char data[1000];
} segment;

void setIP(char *dst, char *src) {
    if(strcmp(src, "0.0.0.0") == 0 || strcmp(src, "local") == 0
            || strcmp(src, "localhost")) {
        sscanf("127.0.0.1", "%s", dst);
    } else {
        sscanf(src, "%s", dst);
    }
}

int main(int argc, char* argv[]){
	int sendersocket;
	struct sockaddr_in sender, agent;
	int ip, port, agentip, agentport;
	char filedata[1000];
	FILE *file;
	if(argc != 4){
        fprintf(stderr,"用法: %s <agent IP> <agent port> <filename>\n", argv[0]);
        fprintf(stderr, "例如: ./sender local 8888 test.txt\n");
        exit(1);
    } 
    else{
    	setIP(agentip, argv[1]);
    	sscanf(argv[2], "%d", &agentport)
    }
	ip = '127.0.0.1';
	port = 8887;
	sendersocket = socket(PF_INET, SOCK_DGRAM, 0);
	file = fopen(argv[3],"r");
	if(file == NULL){
		printf("open failure");
		return 1
	}

	sender.sin_family = AF_INET;
    sender.sin_port = htons(port);
    sender.sin_addr.s_addr = inet_addr(ip);
    memset(sender.sin_zero, '\0', sizeof(sender.sin_zero)); 

    agent.sin_family = AF_INET;
    agent.sin_port = htons(agentport);
    agent.sin_addr.s_addr = inet_addr(agentip);
    memset(agent.sin_zero, '\0', sizeof(agent.sin_zero));  

    bind(sendersocket,(struct sockaddr *)&sender,sizeof(sender));

    agent_size = sizeof(agent);

    printf("agent info: ip = %s port = %d\n", ip[1], port[1]);
    int stop = 0;
    int num = 1;
    while(stop == 0){
    	segment s_tmp;
    	ret = fread(filedata,1,1000,file);
    	if(ret != 1000){
    		len = strlen(filedata);
    		for(int i = len;i < 1000;i ++){
    			filedata[i] = 0;
    		}
    	}
    	else if(ret <= 0){
    		s_tmp.head.fin = 1;
    		for(int i = 0;i < 1000;i ++){
    			filedata[i] = 0
    		}
    	}
        strcpy(s_tmp.data, filedata);
        s_tmp.head.length = sizeof(s_tmp);
        s_tmp.head.seqNumber = num;
        num ++;
        s_tmp.head.fin = 0, s_tmp.head.syn = 0, s_tmp.head.ack = 0;
        sendto(sendersocket, &s_tmp, s_tmp.head.length, 0, (struct sockaddr *)&agent,sizeof(agent));
    }

}